### PR TITLE
Implement the xTB for the TorsionDrive

### PR DESCRIPTION
**Pull request by xiki-tempula**
Though MolSSI/QCEngine allows the usage of TorsionDrive/xTB using the geomeTRIC optimiser, it doesn't allow one to use the native optimiser.
From my experience, the native optimiser is both more stable and faster compared with geomeTRIC.

This PR allows TorsionDrive to use the xTB as the QM engine and using native-opt.
I have also optimised the parallelisation routine so that xTB will be using one thread (faster than using multiple threads), the work queue could then initiate multiple optimisations on the same machine.

This PR also brought a minor change.
Change the absolute import to local import

### CODE REVIEW

**Comment by xiki-tempula** [comment]
@leeping Sorry for disturbing you. I wonder if you mind have a review of this PR, please? Thank you.

**Comment by LilDojd** [comment]
@xiki-tempula Thank you so much for the commits!

I did the tests, and it seems to work properly, if no additional constraints are passed explicitly. 
What I mean by that is the cell energy does not converge if you pass the following  to `self.opt_restrain` in class `EnginexTB`:

> $constraint
>     force constant=1.0
>     atoms: i, j, k

My hypothesis is that xTB by default applies per pair harmonic potentials based on input geometry, which changes as the wavefront is propagated. We could avoid this behavior either by fixing the atoms or applying a stronger constrain, however, there are cases when some flexibility is preferred.

The workaround is to pass `reference=coord.ref`, where `coord.ref` is the first optimized structure

I could work on this and add some tests if this branch is not stale.

**Comment by xiki-tempula** [comment]
@LilDojd Thanks for the review. Sorry I don't quite understand the issue.
So I do know that the dihedral constrain is done in a harmonic restraint fashion but I'm not quite sure of the issue with this approach.

> My hypothesis is that xTB by default applies per pair harmonic potentials based on input geometry, which changes as the wavefront is propagated. We could avoid this behavior either by fixing the atoms or applying a stronger constrain, however, there are cases when some flexibility is preferred.

Do we want to engage xTB developer @awvwgk who might be able to offer more insight?

**Comment by LilDojd** [comment]
I will try to explain the issue in a bit more detail, as this might apply not only to xTB case, but TorsionDrive in general.

The way I understand the current implementation of extra constraints in TorsionDrive, and correct me if I am wrong, is that it uses `$set` and `$freeze` directives of geomeTRIC which basically enforce a value of length/angle/dihedral or cartesian coordinates by setting the gradient of mentioned degrees of freedom to 0. This works just fine in many cases, e.g if you want to freeze a specific rotatable bond in a constrained optimization. However, when working with larger moieties (for example, ligand molecule in its protein pocket environment) you might want to allow some flexibility of the pocket residues. GFN2-xtb allows just that with a `$constrain` directive and setting `force constant` to a smaller value, e.g 0.5 Hartree/Bohr², basically making a REstraint. 

xtb works with z-matrices, so in case of applying positional restraints to specific atoms, it forms a set of pairwise harmonic potentials between restrained atoms with r0 defined by input geometry.

From xtb docs for caffeine molecule:

<pre><code>
           -------------------------------------------------
          |                   Constraints                   |
           -------------------------------------------------

 * 15 constrained atom positions
   <b>positions referring to input geometry</b>

     #   Z                                   position/Å      displ./Å

    11   8 O      6.3577347    -3.6327225    -0.0684681     0.0000000

     1   6 C      1.0687445     0.0520162    -0.0755782     0.0000000
     3   6 C      3.3535252     1.0744217    -0.0774722     0.0000000
     5   6 C      4.5969189    -0.6303196    -0.0745855     0.0000000
     6   6 C      3.2896462    -1.0950551    -0.0735452     0.0000000
     7   6 C      2.9629004    -2.4886091    -0.0702591     0.0000000
    10   6 C      5.4425717    -2.8389078    -0.0699228     0.0000000
    13   6 C      7.0086271    -0.9538835    -0.0749134     0.0000000
    14   6 C      3.9536622    -4.7147069    -0.0641970     0.0000000
     2   7 N      2.5030143     0.0336686    -0.0754494     0.0000000
     4   7 N      4.6213728     0.7205067    -0.0770658     0.0000000
     9   7 N      4.1215924    -3.2704219    -0.0683024     0.0000000
    12   7 N      5.6601563    -1.4769082    -0.0732293     0.0000000
     8   8 O      1.8493654    -2.9780046    -0.0690630     0.0000000

    11   8 O      6.3577347    -3.6327225    -0.0684681     0.0000000


     <b>applying 105 atom pairwise harmonic potentials</b>
       applied force constant per pair:     0.0035714
     effective force constant per atom:     0.0500000
         constraining energy/grad norm:     0.0000000     0.0000000
</code></pre>

Alternatively, one could pass a different reference geometry by `reference=coord.ref` in $constraint

This is a part of my log:

<pre><code>
* 30 constrained atom positions
   <b>positions refering to 'coord.ref'</b>

     #   Z                                   position/Å      displ./Å
     1   6 C      7.1387529     5.5225370    -0.4255289    46.7856280
     2   8 O      6.7322052     5.5486268     0.7358838    46.1751507
     3   6 C      8.4352404     6.2368216    -0.8100071    47.8225271
    48   6 C     -3.6432530     2.5188771    -7.1096075    46.4913725
   . . . . . . . . . . . . .
   302   6 C      1.8326172    -9.6246336     0.6041190    30.7140954
   324   6 C     -4.9089984    -7.3981878    -2.6710114    34.0620439
   334   7 N     -5.7681473    -6.6908796    -1.9376089    34.3915645
   335   6 C     -6.9617571    -7.2763057    -1.3642881    33.5058175

     applying 435 atom pairwise harmonic potentials
       applied force constant per pair:     0.0344828
     effective force constant per atom:     1.0000000
         constraining energy/grad norm:     0.0000000     0.0000182

     total constraint energy/grad norm:     0.0000000     0.0000182
</code></pre>


In TorsionDrive, the active points from last step are used to launch constrained optimizations to neighboring points. If the optimization produces a point with lower energy, it sets the grid cell to active, and so on. This means that each time the restraint potential is recalculated with respect to the structure from previous step, meaning the energy after optimization is always lower, as it drifts to "unrestrained" optimized structure. This produces the behavior observed in bottom image:

Demonstration

So basically, adding weak restraints without a single reference structure just increases the time to convergence, which is obvious after running TorsionDrive on a 200 CPU server for 24 hours. I hope I made this more clear

I am sorry I can not provide the files needed to reproduce this behavior as of right now

I do not think we should involve xTB developer, as this behavior can be accounted for internally as a part of extra constraints

File: torsiondrive/qm_engine.py [hunk 1]
```diff
@@ -327,6 +330,114 @@ def load_native_output(self, filename='output.dat'):
...
+        elems = []
+        with open(input_file, 'r') as xTBin:
+            n_atoms = int(xTBin.readline())
+            comment = xTBin.readline().strip()
+            for line in range(n_atoms):
+                elem, x, y, z = xTBin.readline().split()
+                elems.append(elem)
+                coords.append((x, y, z))
+
+        if comment[:len('xTB arguments:')] == 'xTB arguments:':
```
Comment (awvwgk) [inline torsiondrive/qm_engine.py]:
```suggestion
        if comment.startswith('xTB arguments:'):
```

File: torsiondrive/qm_engine.py [hunk 2]
```diff
@@ -327,6 +330,114 @@ def load_native_output(self, filename='output.dat'):
...
+            cmd = comment[len('xTB arguments:'):]
+        else:
+            cmd = '--opt'
+            print("'xTB arguments:' not found in the command line, no extra arguments passed to xTB.")
+
+        if self.native_opt:
+            self.temp_type = "optimize"
+            assert '--opt' in cmd, "comment line should contain --opt command to use native opt"
```
Comment (awvwgk) [inline torsiondrive/qm_engine.py]: Shouldn't this rather throw than assert?

File: torsiondrive/qm_engine.py [hunk 3]
```diff
@@ -327,6 +330,114 @@ def load_native_output(self, filename='output.dat'):
...
+            outfile.write(self.opt_restrain)
+
+    def optimize_native(self):
+        """ run the constrained optimization using native opt, in 2 steps:
+        1. write a optimization job input file.
+        2. run the job
+        """
+        if self.extra_constraints is not None:
+            raise RuntimeError('extra constraints not supported in xTB native optimizations')
+        # add the optking command
```
Comment (awvwgk) [inline torsiondrive/qm_engine.py:395]: Don't think that has to do anything with `xtb`.

File: torsiondrive/qm_engine.py [hunk 4]
```diff
@@ -327,6 +330,114 @@ def load_native_output(self, filename='output.dat'):
...
+        with open(filename) as outfile:
+            n_atoms = int(outfile.readline())
+            comment = outfile.readline()
+            for line in range(n_atoms):
+                elem, x, y, z = outfile.readline().split()
+                elems.append(elem)
+                coords.append((x, y, z))
+        try:
+            # Also checks if right number of fields are present
+            _, final_energy, _, _, _, _, _ = comment.split()
```
Comment (awvwgk) [inline torsiondrive/qm_engine.py]: This output was changed in the past and is not guaranteed to stay in the current format. The first real number being the energy however is something you can rely on

**Comment by LilDojd** [comment]
I am sorry, I made an error in the image. Obviously, bottom arrow should go to middle grid state

**Review by awvwgk — COMMENTED** [review]
For the dihedral restraint I think we apply

$E = f \cdot \sin[\phi - \phi_0 + \pi]$

The implementation can be found here

https://github.com/grimme-lab/xtb/blob/1922dd31cc734edb136d3ef5ea42c7d542170e8d/src/constr.f90#L322-L325

For the atom-wise restraint we indeed use pairwise quadratic or higher polynomials.

**Comment by leeping** [comment]
My apologies for the delays! I will look at these as soon as possible.

**Comment by selimsami** [comment]
Hi @leeping, @xiki-tempula  , is the xTB implementation ready to be added to the main branch?

File: torsiondrive/qm_engine.py (hunk 3 above)
Comment (xiki-tempula) [reply torsiondrive/qm_engine.py:395]: Added TorsionDrive xTB native optimizations to show that this is not a xTB issue but rather I don't have time to work on it.

**Comment by leeping** [comment]
I'll review the code in the coming days.  One change I recently made on the geomeTRIC side was a workaround to set the charge and spin multiplicity in the ASE interface to xtb-python.  Thanks.

**Comment by leeping** [comment]
I have a few questions:

1) There seem to be a few ways to perform torsion drives using the GFN-xTB types of semiempirical models.  There's two existing ways, and one way that I would like to add:
a) Using geomeTRIC's interface to QCEngine, which has an interface to xTB.
b) Using xTB's native optimizer (this PR).
c) Using geomeTRIC's interface to ASE, which has an interface to xTB (currently being worked on in geomeTRIC).

I would like to enable all three.  One question is, do we all agree (c) should be independent from this PR?  Because even though ASE supports xtb as one of its Calculator classes, it also supports other methods and software packages.

2) The xtb geometry optimizer is different from geomeTRIC and the other QM native optimization methods because it is using an energy restraint rather than an exact constraint.  That is fine, but I don't want to create the mistaken impression that one is better, because an un-constrained optimization with an energy restraint is in some sense an "easier" job.  Could we print a message to indicate that constraints will not be satisfied exactly when using xtb's native optimizer?

3) Is there a use case for requesting more than one core when running the geometry optimizations?  My impression has been that TorsionDrive often creates enough optimization jobs to saturate the number of available cores, and in those cases, it's best to use one core per geometry optimization job.  However, if we allow xTB to use multiple cores in the geometry optimization, we should probably enable this feature for other engines as well in a future PR.

Thanks,

- Lee-Ping

**Comment by jevandezande** [comment]
In my experience with xtb for systems < 500 atoms, speedups from more than 4 cores is marginal, and even at 4 cores the efficiency is pretty low. Additionally, one must set both the parallel flag, `-P`, and the environment variables, `OMP_NUMTHREADS` and `MKL_NUMTHREADS` (https://xtb-docs.readthedocs.io/en/latest/setup.html#parallelisation). Using the inherent parallelism of TorsionDrive jobs is probably best.

I'm personally a fan of getting this PR merged with just option b, as it works well enough for my use cases, and then going back later and adding other methods.

**Comment by xiki-tempula** [comment]
Hi @leeping Thank you for agree to review.

> There seem to be a few ways to perform torsion drives using the GFN-xTB types of semiempirical models. There's two existing ways, and one way that I would like to add:
> a) Using geomeTRIC's interface to QCEngine, which has an interface to xTB.
> b) Using xTB's native optimizer (this PR).
> c) Using geomeTRIC's interface to ASE, which has an interface to xTB (currently being worked on in geomeTRIC).
> I would like to enable all three. One question is, do we all agree (c) should be independent from this PR? Because even though ASE supports xtb as one of its Calculator classes, it also supports other methods and software packages.

I think there is already a solution to `a` in QC engine, which uses xTB for gradient and then use geomtric to do geometry optimisation. https://docs.qcarchive.molssi.org/projects/QCPortal/en/stable/_modules/qcportal/models/torsiondrive.html

I think the main merit of this PR is in terms of speed. Given that xTB is so fast, the speed bottleneck is the IO. So using the native optimiser makes the run much faster as the reading and the writing of the gradient becomes the speedbottle for other methods.

> The xtb geometry optimizer is different from geomeTRIC and the other QM native optimization methods because it is using an energy restraint rather than an exact constraint. That is fine, but I don't want to create the mistaken impression that one is better, because an un-constrained optimization with an energy restraint is in some sense an "easier" job. Could we print a message to indicate that constraints will not be satisfied exactly when using xtb's native optimizer?

This PR hasn't implemented the custom constrain so I think this might be an issue for future PR.  

> Is there a use case for requesting more than one core when running the geometry optimizations? My impression has been that TorsionDrive often creates enough optimization jobs to saturate the number of available cores, and in those cases, it's best to use one core per geometry optimization job. However, if we allow xTB to use multiple cores in the geometry optimization, we should probably enable this feature for other engines as well in a future PR.

Given xTB is so fast, the parallelisation gives performance regression for normal molecules as the overhead of parallisation would dominate the calculation. If we limit the usage of TorsionDrive to small molecules, one core would be the optimal approach.

**Comment by leeping** [comment]
> I think the main merit of this PR is in terms of speed. Given that xTB is so fast, the speed bottleneck is the IO. So using the native optimiser makes the run much faster as the reading and the writing of the gradient becomes the speedbottle for other methods.

I agree. I think there is definitely a benefit to this PR and I plan on merging it after I understand all of the changes and we agree on the requested changes.

>> The xtb geometry optimizer is different from geomeTRIC and the other QM native optimization methods because it is using an energy restraint rather than an exact constraint. That is fine, but I don't want to create the mistaken impression that one is better, because an un-constrained optimization with an energy restraint is in some sense an "easier" job. Could we print a message to indicate that constraints will not be satisfied exactly when using xtb's native optimizer?
> 
> This PR hasn't implemented the custom constrain so I think this might be an issue for future PR.

I'm not actually referring to custom constraints (I call them "extra constraints"), but to xtb's use of energy restraints (i.e. harmonic penalty functions) instead of true constraints that use Lagrange multipliers or other means.  As a result, xtb's geometry optimizations converge faster but do not give exact constraint satisfaction.  I haven't used xtb before, so please let me know if I'm misunderstanding how xtb works.

All I'm asking for right now is for torsiondrive to print a message at the start of a job that uses xtb that notifies the user that energy restraints are used in place of true constraints.

>> Is there a use case for requesting more than one core when running the geometry optimizations? My impression has been that TorsionDrive often creates enough optimization jobs to saturate the number of available cores, and in those cases, it's best to use one core per geometry optimization job. However, if we allow xTB to use multiple cores in the geometry optimization, we should probably enable this feature for other engines as well in a future PR.
> 
> Given xTB is so fast, the parallelisation gives performance regression for normal molecules as the overhead of parallisation would dominate the calculation. If we limit the usage of TorsionDrive to small molecules, one core would be the optimal approach.

Do you see a strong use case to use TorsionDrive/xTB for large molecules?  If so, then we can keep the multiple cores feature and extend it to the other engines.

**Comment by leeping** [comment]
> In my experience with xtb for systems < 500 atoms, speedups from more than 4 cores is marginal, and even at 4 cores the efficiency is pretty low. Additionally, one must set both the parallel flag, `-P`, and the environment variables, `OMP_NUMTHREADS` and `MKL_NUMTHREADS` (https://xtb-docs.readthedocs.io/en/latest/setup.html#parallelisation). Using the inherent parallelism of TorsionDrive jobs is probably best.
> 
> I'm personally a fan of getting this PR merged with just option b, as it works well enough for my use cases, and then going back later and adding other methods.

Thanks a lot - that is my plan as well.

**Comment by leeping** [comment]
I've finished reviewing the codes, and I think everything on the whole looks good.  There's just two changes I'd like to request before merging:

- Please add the printed message about energy restraints when using the xtb engine as mentioned above.  I mainly don't want to give users the impression that geomeTRIC is an inferior geometry optimization method when it is in fact running a different kind of job from xtb's optimizer.
- Please remove the `--core` feature.  I think it will require a more significant effort to enable multi-core geometry optimizations in a way that is consistent across multiple engines, and in many use cases it may actually cause a performance regression.

**Comment by xiki-tempula** [comment]
@leeping Thanks for the review. I have removed the --core option and add a warning when native opt is choosen.

**Comment by leeping** [comment]
Thank you @xiki-tempula - as a final step, are you able to check the coverage on this PR?  I want to ensure that the percentage of code covered by tests does not decrease.

- Lee-Ping

**Comment by xiki-tempula** [comment]
@leeping I have run the test coverage. I'm sorry that I cannot get some test to run such as `test_reproduce_extra_constraints_example` Where I got
`Exception: Could not create queue on port 50124`. But I guess these tests are not related to the code that I have produced here.

So here is the coverage before this PR.

```
Name                                Stmts   Miss  Cover
-------------------------------------------------------
setup.py                                4      4     0%
torsiondrive/__init__.py                5      0   100%
torsiondrive/dihedral_scanner.py      410     17    96%
torsiondrive/extra_constraints.py     109      6    94%
torsiondrive/launch.py                101      8    92%
torsiondrive/priority_queue.py         22      0   100%
torsiondrive/qm_engine.py             433     26    94%
torsiondrive/td_api.py                203     29    86%
torsiondrive/tools/__init__.py         35      0   100%
torsiondrive/tools/plot_1d.py          37      1    97%
torsiondrive/tools/plot_2d.py          57      5    91%
torsiondrive/wq_tools.py               61     56     8%
versioneer.py                         577    577     0%
-------------------------------------------------------
TOTAL                                2054    729    65%
```

This is the the coverage after this PR.

```
Name                                Stmts   Miss  Cover
-------------------------------------------------------
setup.py                                4      4     0%
torsiondrive/__init__.py                5      0   100%
torsiondrive/dihedral_scanner.py      410     17    96%
torsiondrive/extra_constraints.py     109      6    94%
torsiondrive/launch.py                111      8    93%
torsiondrive/priority_queue.py         22      0   100%
torsiondrive/qm_engine.py             499     28    94%
torsiondrive/td_api.py                203     29    86%
torsiondrive/tools/__init__.py         35      0   100%
torsiondrive/tools/plot_1d.py          37      1    97%
torsiondrive/tools/plot_2d.py          57      5    91%
torsiondrive/wq_tools.py               61     56     8%
versioneer.py                         577    577     0%
-------------------------------------------------------
TOTAL                                2130    731    66%
```

**Comment by leeping** [comment]
Thank you so much for your patience.  I will merge the feature now.

- Lee-Ping
